README.md
## [SDA - Sistemas Distribuidos para Automação]

Servidor OPC Gateway para envios de variaveis
do processo via TPC/IP para um host remoto.

- Server.py:  Cliente OPC COM , Servidor TCP/IP

- Cliente.py: Cliente TCP/IP

- Arduino server: OPC Arduino server para ler e
disponiblizar via COM os valores dos items
definidor no arduino.

- arduino-opc: Código do arduino com leitura e
escrita de dois items. Ze1 e Ve1.


ArduinoOPC-Client/changelog.txt
Version 1.9
-----------
- Reduced itemsmap and json response for faster processing

Version 1.8
-----------
- Improved data recovery
- Fast and reliable
- Some bug fixed

Version 1.7
-----------
- Fast recovery of the OPC Items map for Ethernet and YUN
- Some bugs fixed and performance improved
- Examples updated
- Please update the OPC Library from Arduino IDE library manager using the 1.7 version

Version 1.6
-----------
- Support for Arduino Ethernet Shield
- Changed OPC Server Name from ArduinoSerialOPCDAServer.exe to a more generic name: ArduinoOPCServer.exe
- Half executable size
- New sketch for Arduino UNO Ethernet

Version 1.5
-----------
- Now can write/send item values to your Arduno YUN using OPC Server 
- Some bugs fixed and performance improved

Version 1.4
-----------
- Supports for Arduino YUN and WIFI
- Allow to get items using a Internet Browser in JSON format
- Some new examples added with JSON support (now at https://github.com/OPCMakers/OPC )

Version 1.3
-------------------------
- Fixed item value called twice so x2 performance!!!
- Better memory management with less footprint
- Writevalue fixed
- Fixed opc items data type
- Float support
- Some new examples added (now at https://github.com/OPCMakers/ArduinoOPCServer )
- All Serial port properties are available in ini file to fine tunning
- Beta library for Arduino Ethernet (not working yet but some code added to unify serial and ethernet library)

Version 1.2 (beta)
------------------
- MultiArduino support!!!!
- Bug fixes
- Sample sketches updated
- Browse items timeout increased
- Add Arduino icon added
- Log info with OPCServer application path to avoid  call old OPCServers
- Added some more log information for debug pourpose
- Now you can change the serial control flow
- Added vertical scroll bar to the debug window


Version 1.1
-----------
- Ability to write boolean values from OPC client
- Port configuration shown on main form
- Changed name to Arduino Serial OPC DA Server
- Added example sketches with the library
- Version info shown in main form
- Bugs fixed and improved performance
- TrayIcon


Version 1.0
-----------
- First relase
- Arduino Uno Tested
- Boolean, byte and int opc item type supported
- Callback opc item values recovery
- Included OPC.h library with small footprint

Client/client.py
#socket_echo_client.py
import pywintypes
import socket
import sys
pywintypes.datetime = pywintypes.TimeType

def get_ip_address():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(("8.8.8.8", 80))
    return s.getsockname()[0]
get_ip_address

ip_gatway = "localhost"

#Interface Principal Client:
print("Client TCP/IP-OPC:")
print("\n Digite o IP do GATEWAY OPC: ")
ip_gatway = input()

if ip_gatway == "":
    ip_gatway = get_ip_address()

# Create a TCP/IP socket
sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)

# Connect the socket to the port where the server is listening
server_address = (ip_gatway, 10000)
print('connecting to {} port {}'.format(*server_address))
sock.connect(server_address)

try:

    


    while True:
        # Send message requiring arduino data
        message = b'Send data'
        sock.sendall(message)
        amount_received = 0
        # Look for the response
        data = sock.recv(100000)
        print('received:  {!r}'.format(data))

finally:
    print('closing socket')
    sock.close()

Server/server.py
#socket_echo_server.py

import time
import pywintypes
import OpenOPC
import socket
import sys

pywintypes.datetime = pywintypes.TimeType
opc = OpenOPC.client()

def startOpcClient():
    opc.connect(opc.servers()[0],'localhost')
    print("Conectado ao servidor OPC: ")
    print(opc.properties('ArduinoSerial0.Ze.Int1'))
startOpcClient


def get_ip_address():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    s.connect(("8.8.8.8", 80))
    return s.getsockname()[0]
get_ip_address


#Interface Principal:
print("Gateway OPC:")
print(" IP: " +  get_ip_address())
print(" OPC Servers Dispniveis: " +"".join(str(x) for x in opc.servers()))
print("\n Digite <ENTER> para conectar ao servidor opc e iniciar o gateway ou <S> para sair:")
keyboard_in = ''
keyboard_in = input()
while True:
    if(keyboard_in == ""):
        startOpcClient()
        break
    elif(keyboard_in == ('s' or 'S')):
        print("Saindo...")
        break
    else:
        keyboard_in = input("Entrada invalida digite novamente:")


#Create a TCP/IP socket
sock = socket.socket(socket.AF_INET,socket.SOCK_STREAM)

#Bind the socket to the port
server_adress = (get_ip_address(),10000)
print('starting up on {} port {}'.format(*server_adress))
sock.bind(server_adress)

#Listen for incoming connections
sock.listen(1)

while True:
    #Wait for a connection
    print('waiting for a connection')
    connection, client_address = sock.accept()
    try:
        print('connection from',client_address)


        #Receive message to send arduino data
        while True:
            data = connection.recv(255)
            if data:
                TAG = opc.properties('ArduinoSerial0.Ze.Int1')
                name = str(TAG[0][2])
                valuea = str(TAG[2])
                message = name + valuea
                value = str.encode(message)
                print('Enviando estado atual do arduino...')       
                #while True:    
                connection.sendall(value)             
            else:
                print('no data from',client_address)
                break
    finally:
        #Clean up the connection
        connection.close()